core/node/node_framework/src/service/named_future.rs

```diff
@@ -0,0 +1,51 @@
+use std::{fmt, future::Future, pin::Pin, task};
+
+use pin_project_lite::pin_project;
+
+use crate::task::TaskId;
+
+pin_project! {
+    /// Implements a future with the name tag attached.
+    pub struct NamedFuture<F> {
+        #[pin]
+        inner: F,
+        name: TaskId,
+    }
+}
```

Comment:
Probably name propagation can be implemented in a cleaner way, it was the simplest option I could do quickly (and IMO still better than what we had before).